### ISSUE TITLE
ODE solver fails with 'The maximum number of solver steps was reached. Try increasing `max_steps`'

### ISSUE THREAD

**Issue by PeishiJiang** [patrick-kidger/diffrax#207]
Hi,

I was playing with this cool package on a chemical reaction ODE problem. This problem solves the time evolution of seven chemical concentrations, which is a stiff problem but can be solved using a Fortran-based solver. However, the diffrax version fails, with an XlaRuntimeError complaining 'The maximum number of solver steps was reached. Try increasing `max_steps`'. Unfortunately, the error persists no matter how large the `max_steps` is and which solver is used (e.g., impliciteuler or Kvaerno5). Note that when commenting the error message in `diffeqsolve` function, I find that the code can solve about the first 100s and output inf (from `solution.ys`) in a later time.

Any suggestion would be appreciated!

Below is the code snippet -- 

```python
from diffrax import diffeqsolve, ODETerm, SaveAt
from diffrax import NewtonNonlinearSolver, Dopri5, Kvaerno3, ImplicitEuler, Euler, Kvaerno5
from diffrax import PIDController

import jax
import jax.numpy as jnp
import jax.random as jrandom

from jax.config import config
config.update("jax_enable_x64", True)

def funclog2(t, logy, args):
    k1, k2, k3 = args[0], args[1], args[2]
    kd1, kd2, kd3 = args[3], args[4], args[5]
    ka1, ka2, ka3 = args[6], args[7], args[8]
    r4 = args[9]
    
    y = jnp.power(10, logy)
    doc, o2, no3, no2, n2, co2, bm = y
    
    # log transform scale
    scale = 1 / jnp.log(10)
    scale = scale / y
    
    # The stoichiometry matrix
    stoich = jnp.array([
        [-1, -1, -1, 5],
        [0, 0, -1, 0],
        [-2, 0, 0, 0],
        [1, -1, 0, 0],
        [0, 1, 0, 0],
        [1, 1, 1, 0],
        [0, 0, 0, -1]
    ])
    
    # Scale stoich
    stoich = jax.vmap(lambda a, b: a*b, in_axes=0)(scale, stoich)
    
    # Reaction rate
    r1 = k1 * bm * doc/(doc+kd1) * no3/(no3+ka1)
    r2 = k2 * bm * doc/(doc+kd2) * no2/(no2+ka2)
    r3 = k3 * bm * doc/(doc+kd3) * o2/(no2+ka3)
    
    r = jnp.array([r1, r2, r3, r4]).T
    
    return stoich @ r

# Static parameters
k1, k2, k3 = 3.24e-4, 2.69e-4, 9e-4 # [mol/L/sec/mass [BM]]
kd1, kd2, kd3 = 2.5e-4, 2.5e-4, 2.5e-4 # [mol/L]
ka1, ka2, ka3 = 1e-6, 4e-6, 1e-6  # [mol/L]
r4 = 2.8e-6 # [mol/L/sec]
args = jnp.array([k1, k2, k3, kd1, kd2, kd3, ka1, ka2, ka3, r4])

# The initial concentrations with the following order [mol/L]:
# doc, o2, no3, no2, n2, co2, bm
# y0 = jnp.array([4.16e-05, 0.000266, 0.000396, 1e-10, 1e-10, 0.00248, 0.0003])
y0 = jnp.array([4.16e-05, 0.000266, 0.000396, 1e-3, 1e-3, 0.00248, 0.0003])
logy0 = jnp.log10(y0)

term = ODETerm(funclog2)
# solver = Dopri5()
# solver = Euler()
solver = Kvaerno5(NewtonNonlinearSolver(rtol=1e-3, atol=1e-6))
# solver = ImplicitEuler(NewtonNonlinearSolver(rtol=1e-3, atol=1e-6))
# t0, t1, dt0 = 0, 3600*24*30, 1
t0, t1, dt0 = 0, 200, 0.01
# t0, t1, dt0 = 0, 3600*24, 3600
solution = diffeqsolve(term, solver, t0=t0, t1=t1, dt0=dt0, max_steps=400000,
                       stepsize_controller=PIDController(rtol=1e-3, atol=1e-6),
                       saveat = SaveAt(t0=True, ts=jnp.linspace(t0,t1)), 
                       y0=logy0, args=args)
solution.stats
```

**Comment by jaschau** [comment]
Hi, are you running on GPU or on CPU? In my experience, numerical stability varies from GPU to GPU and stability is highest on CPU. So you might want to try switching to CPU and see if the issue persists. Another thing to try that is also mentioned in the FAQ https://docs.kidger.site/diffrax/further_details/faq/ would be switching from float32 accuracy (which is the default in Jax) to float64.

**Comment by patrick-kidger** [comment]
I've had a quick look at this. By operating in logspace I think you're doing something pretty numerically unstable.

(Certainly I've had issues with such biological models in logspace before, due to the numerics breaking down.)

Running on the CPU, and modifying your vector field to:
```python
from equinox.internal import error_if

def funclog2(t, logy, args):
    logy = error_if(logy, jnp.invert(jnp.isfinite(logy)), "kaboom1")
    ...   # as before
    out = stoich @ r
    jax.debug.print("logy: {}, out: {}", logy, out)
    out = error_if(out, jnp.invert(jnp.isfinite(out)), "kaboom2")
    return out
```
Then I get `kaboom2`. That is, the vector field was passed valid values and returned invalid values.
In practice, looking at the numbers printed by `jax.debug.print`, I see that the last few steps were
```
logy: [-2.76467277 -3.58396695 -3.41860059 -2.99945631 -2.99736787 -2.60226884                                                                                          
 -0.56311583], out: [-4.62254990e-02 -9.31044518e-02 -1.75688609e-01  5.71850521e-03                                                                                    
  2.76117393e-02  3.42369211e-02 -4.44690832e-06]                                                                                                                       
logy: [ -8.35717563 -16.94683811 -28.68546674  -2.17873029   0.96463305                                                                                                 
   2.3083941  134.5808707 ], out: [-1.77908854e+134 -3.94856242e+128 -1.88411214e+132 -1.17964744e+128                                                                  
  8.47986396e+124  3.84262354e+123 -3.19206440e-141]                                                                                                                    
logy: [-3.60147931e+135  6.31517985e+131 -4.67530871e+133 -4.24622231e+130                                                                                              
 -1.90710947e+131 -2.37392278e+131  4.73345703e+002], out: [nan nan nan nan nan nan nan]
```
working through this, we see:
1. the last normal step
2. a step in which `logy` has reasonable values but `out` has extreme values.
3. a step in which `logy` has extreme values and `out` has NaNs.

On a step in which the vector field outputs a NaN, then Diffrax will keep retrying with smaller and smaller step sizes. Possibly what's going on here is that we're looking at the vector field evaluations across the start/end of a step, and so `logy` has already taken on the extreme-but-not-invalid values.

---

We can first try and avoid the above issue by taking smaller, more careful steps. (Your tolerances are very large!) Switching to `rtol=1e-12` and `atol=1e-12`, and adding `t` to the above `jax.debug.print`, I can see that the solve stalls at about `t=107.14285715882677`.

This seems to indicate that your ODE is exploding in finite time.

[If you haven't seen explosion in finite time before, then consider e.g. `y(0) = 1; dy/dt = -y(t)^2`. This has solution `y(t) = 1 / (1 - t)`, which blows up as `t -> 1`. Note that the vector field f(y) = -y^2 is not Lipschitz.]

Indeed, I note that your vector field is non-Lipschitz (you divide by a state variable), so the conditions of [Picard's Existence Theorem](https://en.wikipedia.org/wiki/Picard%E2%80%93Lindel%C3%B6f_theorem) aren't satisfied -- i.e. mathematically I don't know of a guarantee that your solution even exists.

And indeed, throwing the above at scipy's Radau solver (which is the the scipy-available solver that is capable of handling the greatest stiffness):
```python
scipy.integrate.solve_ivp(jax.jit(funclog2), [t0, t1], logy0, args=(args,), rtol=1e-12, atol=1e-12, method="Radau")
```
then we once again get that the ODE stalls at `t=107.14285714283885`. (Curious tangent: this seems to be exactly `107 + (1/7)`.)

---

So in summary: your ODE isn't defined past `t = 107 + (1/7)`, but you've asked to solve it up to `t=200`, and this isn't mathematically possible.

**Comment by patrick-kidger** [comment]
Side note, a few ways you can use Diffrax slightly more efficiently:
- You don't need to wrap up your constants into an array here:
    ```
    k1, k2, k3 = 3.24e-4, 2.69e-4, 9e-4 # [mol/L/sec/mass [BM]]
    kd1, kd2, kd3 = 2.5e-4, 2.5e-4, 2.5e-4 # [mol/L]
    ka1, ka2, ka3 = 1e-6, 4e-6, 1e-6  # [mol/L]
    r4 = 2.8e-6 # [mol/L/sec]
    args = jnp.array([k1, k2, k3, kd1, kd2, kd3, ka1, ka2, ka3, r4])
  ```
    as `args` can be PyTree.
- You pass `SaveAt(t0=True, ts=jnp.linspace(t0, t1))`. This will result in `t0` getting recorded twice.

**Comment by PeishiJiang** [comment]
Thank you so much for digging into my codes and explaining the debugging details to me @patrick-kidger ! These are super helpful (particularly on the tips of using jax.debug and equinox.internal.error_if)!

After taking a second look at the code, I did find the bugs that cause the numerical errors. Some calculations of reaction rates (in the vector field function) should be
```python
r3 = k3 * bm * doc/(doc+kd3) * o2/(o2+ka3)
r4 = r4 * bm
```
Once that is fixed, the code runs well.

Regarding the numerical stability, I was performing the ODE on the logarithm of the concentrations (i.e., `logy`) due to the extremely small values of these chemical concentrations, which are sort of numerically unsolvable. Therefore, there is a log scale transformation `scale = 1 / jnp.log(10); scale = scale / y`. Based on your comment, the solution might still be non-exist given an ``ill-posed'' initial condition `logy0`. Is there any other way to resolve this issue?

**Comment by PeishiJiang** [comment]
Thank you for the advice @jaschau. I was running the code on CPU only. @patrick-kidger has kindly helped resolve the issue.
